Docs: Add missing markup to Argument Clinic docs

erlend-aasland opened the pull request:
- Add :py:func:, :c:func:, etc. where appropriate
- Remove repeated links within the same paragraph


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--106876.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

erlend-aasland committed:
Docs: Add missing markup to Argument Clinic docs

- Add :py:func:, :c:func:, etc. where appropriate
- Remove repeated links within the same paragraph

erlend-aasland committed:
Add missing colon

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -162,9 +163,9 @@ Let's dive in!
    or :c:func:`PyArg_ParseTupleAndKeywords`, and hasn't been converted
    to work with Argument Clinic yet.
-   For my example I'm using ``_pickle.Pickler.dump()``.
+   For my example I'm using :py:meth:`!_pickle.Pickler.dump`.
```
I'm probably missing some context, but is there any reason why it's `_pickle` and not `pickle`?
Would it be useful to do something like:
```suggestion
   For my example I'm using :py:meth:`_pickle.Pickler.dump <pickle.Pickler.dump>`.
```
or even just
```suggestion
   For my example I'm using :py:meth:`_pickle.Pickler.dump <pickle.dump>`.
```
?

erlend-aasland replied:
It's `_pickle` because that's the C implementation behind `pickle`. I think it makes sense to link to `pickle.Pickler.dump` anyways. Thanks!

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -392,7 +393,7 @@ Let's dive in!
         [clinic start generated code]*/
 
-12. Save and close the file, then run ``Tools/clinic/clinic.py`` on
+12. Save and close the file, then run :program:`Tools/clinic/clinic.py` on
```
```suggestion
12. Save and close the file, then run :source:`Tools/clinic/clinic.py` on
```
Why not `:source:` like above?

erlend-aasland replied:
I think my reasoning was that the instructions tell you to run `clinic.py` as you'd run an executable, hence the `:program:` role. But, I'm not sure my reasoning holds water. Perhaps this is better:

```suggestion
12. Save and close :source:`Modules/_pickle.c`, then run :program:`python3 ./Tools/clinic/clinic.py Modules/_pickle.c` on
```

erlend-aasland replied:
Alternatively:

```suggestion
12. Save and close the :source:`Modules/_pickle.c`, then run :program:`make clinic` to regenerate
```

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -546,15 +547,15 @@ Let's dive in!
         &_Py_ID(new_unique_py_id)
 
-    If it does, you'll have to run ``Tools/scripts/generate_global_objects.py``
+    If it does, you'll have to run :program:`Tools/scripts/generate_global_objects.py`
```
```suggestion
    If it does, you'll have to run :source:`Tools/scripts/generate_global_objects.py`
```
Same.

erlend-aasland replied:
Some alternatives:

```suggestion
    If it does, you'll have to run :program:`python3 ./Tools/scripts/generate_global_objects.py`
```

or:

```suggestion
    If it does, you'll have to run :program:`make regen-global-objects`
```

erlend-aasland replied:
I resolved most of these remarks in 971c0584fcb7da8809bd0da854963bdf4f5befcd. Hope that looks ok with you. If not, ping me again.

ezio-melotti replied:
The advantage of using `:source:` is that it creates a link that might be interesting for people that want to take a look at the code before running it, whereas `:program:` doesn't.  The docs for `:program"` also say it's for "The name of an executable program.", without mentioning commands you have to execute (that might include paths and cmdline args).

Another thing to keep in mind is how the output looks like:
* [Tools/clinic/clinic.py](https://github.com/python/cpython/tree/3.12/Tools/clinic/clinic.py) -- with `:source:` it looks like a link
* **Tools/clinic/clinic.py** -- with `:program:` it looks bold
* ``Tools/clinic/clinic.py`` -- with ` ``...`` ` it looks monospaced and with a different background

I'm actually not a big fan of how `:program:` is rendered, since it looks just like emphasis rather than a command you have to run.  `:source:` is good when talking about the script in general since it creates the link and you don't have to copy-paste it as a command, but for cases where you do (like the ``make regen-global-objects``) I'm thinking that a simple ` ``...`` ` might be a better option (since it would be more akin to an expression like ``var = func(args)``).

erlend-aasland replied:
> I'm thinking that a simple `` ``...`` `` might be a better option (since it would be more akin to an expression like `var = func(args)`).

I agree; neither `:program:` nor `:source:` are perfect markups here.

erlend-aasland replied:
Reverted in 47061df9dd03fb10c300f8c67d0d9121948ab79f

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -453,11 +454,11 @@ Let's dive in!
     If any of these items differ in *any way*,
     adjust your Argument Clinic function specification and rerun
-    ``Tools/clinic/clinic.py`` until they *are* the same.
+    :program:`Tools/clinic/clinic.py` until they *are* the same.
```
```suggestion
    :source:`Tools/clinic/clinic.py` until they *are* the same.
```

erlend-aasland replied:
... or:

```suggestion
    :program:`python3 Tools/clinic/clinic.py Modules/_pickle.c` until they *are* the same.
```

erlend-aasland replied:
Ah, or perhaps this is the best alternative:

```suggestion
    :program:`make clinic` until they *are* the same.
```

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -576,15 +577,15 @@ Argument Clinic will use that function name for the base (generated) function,
 
 For example, if we wanted to rename the C function names generated for
-``pickle.Pickler.dump``, it'd look like this::
+:py:meth:`!pickle.Pickler.dump`, it'd look like this::
```
Still missing context, but here there's no `_`, so the link should work if you remove the `!` -- unless it's just an example and no link is needed.

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -1100,7 +1101,7 @@ For all of these, return ``-1`` to indicate error.
 To see all the return converters Argument Clinic supports, along with
 their parameters (if any),
-just run ``Tools/clinic/clinic.py --converters`` for the full list.
+just run :program:`Tools/clinic/clinic.py --converters` for the full list.
```
Here it's a bit trickier because you have an arg.  Depending on what you decided to do with the other `:program:`s, applying this change, sticking to ` ``...`` ` or using something like
```suggestion
just run :source:`Tools/clinic/clinic.py --converters <Tools/clinic/clinic.py>` for the full list.
```
might all be valid options.

erlend-aasland replied:
or:

```suggestion
just run :program:`python3 ./Tools/clinic/clinic.py --converters` for the full list.
```

erlend-aasland replied:
I'm not sure `:source:` is better. For now, I've kept this.

erlend-aasland replied:
Reverted in 47061df9dd03fb10c300f8c67d0d9121948ab79f

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -1267,16 +1268,17 @@ module state::
-methods, which cannot use the ``METH_METHOD`` convention.
+The ``defining_class`` converter is not compatible with
+:py:meth:`!__init__` and :py:meth:`!__new__`
+methods, which cannot use the :c:data:`METH_METHOD` convention.
```
```suggestion
methods, which cannot use the :c:data:`!METH_METHOD` convention.
```
I guess this needs a `!` to silence the warning, unless it's an actual target and the warning can be fixed in some other way.

serhiy-storchaka replied:
I think that the right fix of this issue is to add a reference target for METH_METHOD, so this reference would work. Adding `!` to silence a warning will also prevent rendering a reference when a target be fixed.

There is a declaration
```rst
.. data:: METH_METHOD | METH_FASTCALL | METH_KEYWORDS
```
in Doc/c-api/structures.rst, and it is the only definition of METH_METHOD. Can we add ```.. data:: METH_METHOD``` hidden from the output?

erlend-aasland replied:
I agree with Serhiy; we should fix the `METH_METHOD` target first. I'll get to it.

serhiy-storchaka replied:
I am working on this. It is complicated.

erlend-aasland replied:
For now, I've disabled the `METH_METHOD` link: b155e3862df036ca4559290e35844db70e2616b9

erlend-aasland replied:
Resolved with 1a30f3b661622d07ff29b0aafc8dda4546c5a999

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -1389,18 +1391,18 @@ Writing a custom return converter is much like writing
 -------------------------------------------------------
 
-To convert a function using ``METH_O``, make sure the function's
+To convert a function using :c:data:`METH_O`, make sure the function's
```
```suggestion
To convert a function using :c:data:`!METH_O`, make sure the function's
```

erlend-aasland replied:
See https://github.com/python/cpython/pull/106876/files#r1267881674. I think we should add the `METH_O` target first.

erlend-aasland replied:
Resolved with 8f239567112a90ede907e3131e95da89a229efbb

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -1412,24 +1414,25 @@ arguments as positional-only::
 
 
-To convert a function using ``METH_NOARGS``, just don't specify
+To convert a function using :c:data:`METH_NOARGS`, just don't specify
```
```suggestion
To convert a function using :c:data:`!METH_NOARGS`, just don't specify
```

serhiy-storchaka replied:
I think that
```rst
.. data:: METH_NOARGS
```
should be fixed as
```rst
.. c:data:: METH_NOARGS
```

ezio-melotti commented on Doc/howto/clinic.rst:
```diff
@@ -1412,24 +1414,25 @@ arguments as positional-only::
 
 You can still use a self converter, a return converter, and specify
-a ``type`` argument to the object converter for ``METH_O``.
+a *type* argument to the object converter for :c:data:`METH_O`.
```
```suggestion
a *type* argument to the object converter for :c:data:`!METH_O`.
```

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -162,9 +163,9 @@ Let's dive in!
+   For my example I'm using :py:meth:`!_pickle.Pickler.dump`.
 
-2. If the call to the ``PyArg_Parse`` function uses any of the
+2. If the call to the :c:func:`!PyArg_Parse*` function uses any of the
```
Visual difference -- `:c:func:` adds `()` at the end. I think it is OK.

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -181,7 +182,7 @@ Let's dive in!
+   Also, if the function has multiple calls to :c:func:`!PyArg_ParseTuple`
    or :c:func:`PyArg_ParseTupleAndKeywords` where it supports different
    types for the same argument, or if the function uses something besides
    PyArg_Parse functions to parse its arguments, it probably
```
You can use ``` :c:func:`!PyArg_Parse*` ``` here.

erlend-aasland replied:
```suggestion
   :c:func:`!PyArg_Parse*` functions to parse its arguments, it probably
```

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -881,13 +882,13 @@ negative numbers.  You just can't do that with a legacy converter!
 --------------------------------------
 
-When using the ``Py_buffer`` converter
+When using the :c:data:`Py_buffer` converter
```
`Py_buffer` here is not a name of the C variable or constant, but the name of the AC converter. It does not belong neither to the nor to the Python namespaces.

erlend-aasland replied:
True, good observation. Thanks!

erlend-aasland replied:
```suggestion
When using the ``Py_buffer`` converter
```

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -910,12 +911,12 @@ or ``encoding`` (for all the format units that start with ``e``).
 One possible problem with using Argument Clinic: it takes away some possible
 flexibility for the format units starting with ``e``.  When writing a
-``PyArg_Parse`` call by hand, you could theoretically decide at runtime what
+:c:func:`PyArg_Parse` call by hand, you could theoretically decide at runtime what
```
I think it means :c:func:`!PyArg_Parse*`.

erlend-aasland replied:
```suggestion
:c:func:`!PyArg_Parse*` call by hand, you could theoretically decide at runtime what
```

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -910,12 +911,12 @@ or ``encoding`` (for all the format units that start with ``e``).
 flexibility for the format units starting with ``e``.  When writing a
-``PyArg_Parse`` call by hand, you could theoretically decide at runtime what
+:c:func:`PyArg_Parse` call by hand, you could theoretically decide at runtime what
 encoding string to pass in to :c:func:`PyArg_ParseTuple`.   But now this string must
```
This is not specific to `PyArg_ParseTuple()`, but to all `PyArg_Parse*()` functions, and it refers to the same function mentioned at the beginning of sentence. Maybe write it as "to it" or remove it at all?

erlend-aasland replied:
I'll try to rewrite this in a better way.

erlend-aasland replied:
Does f5bb4a318f674840152338b4358b9fe956511ec4 look ok to you?

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -969,7 +970,7 @@ expression.  Currently the following are explicitly supported:
 * String constants
 * ``True``, ``False``, and ``None``
-* Simple symbolic constants like ``sys.maxsize``, which must
+* Simple symbolic constants like :py:attr:`sys.maxsize`, which must
```
In some occurrences it is written as ``` :attr:`sys.maxsize` ```, and in others as ``` :data:`sys.maxsize` ```. What role is more appropriate?

erlend-aasland replied:
I'll investigate. I guess one of the Sphinx experts will know. Cc. @AA-Turner

erlend-aasland replied:
It looks to me like attr is correct in this case.

erlend-aasland replied:
Actually, it is annotated with `.. data::` in `Doc/library/sys.rst`, so we should probably use `py:data:` for `sys.maxsize`.

serhiy-storchaka replied:
`:data:` is purposed for module level variables. Opened #106892 for more global changes.

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -990,28 +991,28 @@ Consider the following example:
 
 If the symbol isn't found in the current module, it fails over to looking in
-``sys.modules``.  That's how it can find ``sys.maxsize`` for example.  (Since you
+:py:attr:`sys.modules`.  That's how it can find :py:attr:`sys.maxsize` for example.  (Since you
```
Same as above, both roles are used with `sys.modules`.

erlend-aasland replied:
`sys.modules` is annotated with `.. data::` in `Doc/library/sys.rst`, so let's go with `:py:data:`

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -990,28 +991,28 @@ Consider the following example:
 the correct equivalent C default value.  So you need to tell it explicitly.
 When you use an expression, you must also specify the equivalent expression
-in C, using the ``c_default`` parameter to the converter:
+in C, using the *c_default* parameter to the converter:
```
In other places the bare ``` `` ``` markup is used for the AC converter parameters: ``` ``subclass_of`` ```, ``` ``type`` ```. I think it should be more consistent.

erlend-aasland replied:
Parameters are usually marked up with pairs of \*, both for C functions and Python functions/methods. I think we should use the same markup for Argument Clinic converters as well, since they are in practice Python functions.

serhiy-storchaka replied:
Then please update also for other parameters, like `subclass_of`.

erlend-aasland replied:
I've marked up what I could find in d96fdebe8f252464eac9bfa6582eed29c54f9f30. I might have missed some, though.

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -1305,36 +1307,36 @@ will be automatically registered with Argument Clinic; its name
     If this is a pointer type, the type string should end with ``' *'``.
 
-``default``
+:py:attr:`!default`
```
Why not the following?
```rst
.. attr:: default
```
You need to declare that the current class is `clinic.CConverter`.

erlend-aasland replied:
Ah, of course.

serhiy-storchaka reviewed (commented):
LGTM in general.

Be aware, there is the concrete function `PyArg_Parse()`, and there is a family of functions `PyArg_Parse*()`. This document may use the former in the meaning of the latter.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address reviews

- line 166: link to pickle.Pickler.dump
- line 187: markup PyArg_Parse*
- line 580: link to pickle.Pickler.dump
- line 891: revert Py_buffer change
- line 919: markup PyArg_Parse
- line 973: markup sys.maxsize as :data:
- line 1008: markup sys.modules as :data:
- markup CConverter and it's attributes as a proper Python class

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Use :data: for METH_O and METH_NOARGS

erlend-aasland committed:
For now, disable the METH_METHOD link

erlend-aasland committed:
Suggest 'make clinic' and 'make regen-global-objects' iso. running clinic directly

erlend-aasland committed:
Markup file suffixes and file paths with :file:

erlend-aasland committed:
Address review: rewrite line 919 similar to suggestion by Serhiy

erlend-aasland committed:
fixup! Suggest 'make clinic' and 'make regen-global-objects' iso. running clinic directly

erlend-aasland committed:
Address review: markup more params with **

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address review: revert :program: markups

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -1137,7 +1140,7 @@ Here's the syntax for cloning a function::
 
 (The functions can be in different modules or classes.  I wrote
-``module.class`` in the sample just to illustrate that you must
+:py:class:`!module.class` in the sample just to illustrate that you must
```
I think that it is rather the AC code than the Python code.

erlend-aasland replied:
The example is for a fictive module called *module* with a class called *class*. I guess it comes down to how you look at it: When you we refer to `_pickle.Pickler.dump` in the other examples, do we view it as part of the Argument Clinic input and use double backticks, or do we view it as a fully qualified named of a Python method thus with `:py:meth:` markup? I definitely view it as the latter. So, IMO, it makes sense to mark up the fictive module and class using `:py:class:!`.

erlend-aasland replied:
After reading it again, I now agree with you. I've reverted the markup change. Thanks for noticing this.

erlend-aasland replied:
See 499ed66666be713b678fa465e24082e92ef61fc5

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -1305,61 +1309,74 @@ will be automatically registered with Argument Clinic; its name
-    Or ``None`` if there is no default.
+      The C type to use for this variable.
+      :attr:`!type` should be a Python string specifying the type,
+      e.g. :class:`int`.
```
```suggestion
      e.g. ``'int'``.
```

erlend-aasland replied:
Thanks! Resolved with fe52717c58828309065f94e360ed6f6dfe7bb2e3

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -1389,18 +1406,18 @@ Writing a custom return converter is much like writing
 -------------------------------------------------------
 
-To convert a function using ``METH_O``, make sure the function's
+To convert a function using :data:`METH_O`, make sure the function's
```
Use `:c:macro:` for `METH_*` constants.

erlend-aasland replied:
Thanks (and thanks for fixing these recently!). Resolved with 1a30f3b661622d07ff29b0aafc8dda4546c5a999

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -392,9 +394,10 @@ Let's dive in!
-12. Save and close the file, then run ``Tools/clinic/clinic.py`` on
-    it.  With luck everything worked---your block now has output, and
-    a ``.c.h`` file has been generated! Reopen the file in your
+12. Save and close the file, then run ``make clinic``
```
Side note: running the script on the specified file and running `make clinic` is not the same. The latter may be more correct, but it is slower.

erlend-aasland replied:
I know, but generally it's better if contributors use `make clinic`, so I think it makes sense to use that throughout the tutorial. I think the "basic usage" section could be improved to mention how to run clinic directly on a specified file.

erlend-aasland replied:
I decided to revert these changes as well; we can change it to the Make target in a future PR.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address review: fix 'int' markup around line 1330

erlend-aasland committed:
Address review: use new and correct directives for C constants/macros

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Revert module.class markup

erlend-aasland committed:
Revert 'make clinic' changes

serhiy-storchaka commented on Doc/howto/clinic.rst:
```diff
@@ -1290,9 +1289,9 @@ after ``self``, or, if ``self`` is not used, as the first argument.  The argumen
-:py:meth:`!__init__` and :py:meth:`!__new__`
-methods, which cannot use the :data:`!METH_METHOD` convention.
+The ``defining_class`` converter is not compatible with :py:meth:`!__init__`
+and :py:meth:`!__new__` methods, which cannot use the :ref:`METH_METHOD
```
BTW you can now use ``` :c:macro:`METH_METHOD` ``` if you prefer.

erlend-aasland replied:
Oh, that's nice; thanks! I'll apply that before landing.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
METH_METHOD now has its own anchor
